v2/workloadapi/bundlesource.go

```diff
@@ -98,31 +133,22 @@ func (s *BundleSource) setX509Context(x509Context *X509Context) {
 
 	newBundles := x509Context.Bundles.Bundles()
 
-	// Add/replace the X.509 content in the SPIFFE bundles. Track the trust
-	// domains represented in the new X.509 context so we can determine if
-	// which existing bundles need their X.509 content removed.
+	// Add/replace the X.509 authorities from the X.509 context. Track the trust
+	// domains represented in the new X.509 context so we can determine which
+	// existing trust domains are no longer represented.
 	trustDomains := make(map[spiffeid.TrustDomain]struct{}, len(newBundles))
 	for _, newBundle := range newBundles {
 		trustDomains[newBundle.TrustDomain()] = struct{}{}
-		existingBundle, ok := s.bundles.Get(newBundle.TrustDomain())
-		if !ok {
-			s.bundles.Add(spiffebundle.FromX509Bundle(newBundle))
-			continue
-		}
-		existingBundle.SetX509Authorities(newBundle.X509Authorities())
+		s.x509Authorities[newBundle.TrustDomain()] = newBundle.X509Authorities()
 	}
 
-	// Remove the X.509 content from bundles that are no longer returned
-	// with the X.509 context. If the bundle is then empty, remove the bundle
-	// from the set.
-	for _, existingBundle := range s.bundles.Bundles() {
-		if _, ok := trustDomains[existingBundle.TrustDomain()]; ok {
+	// Remove the X.509 authority entries for trust domains no longer
+	// represented in the X.509 context.
+	for existingTD := range s.x509Authorities {
+		if _, ok := trustDomains[existingTD]; ok {
 			continue
 		}
-		existingBundle.SetX509Authorities(nil)
-		if existingBundle.Empty() {
-			s.bundles.Remove(existingBundle.TrustDomain())
-		}
+		delete(s.x509Authorities, existingTD)
```

Comment:
This should be protected by a lock.

Reply:
It is. See the top of the setX509Context function.